specs/protocol/safe-extensions-v2.md

```diff
@@ -0,0 +1,489 @@
+
+If an attacker has joint or temporary key control over a quorum of keys, honest users should always be able to recover
+the account.
+
+#### Severity: Critical
+
+If this invariant is broken, honest control of the multisig is lost.
+
+### iSS-008: Owners Of Child Safes Can Signal Rejection Of Transactions
+
+A nested safe setup is a safe (parent safe) in which one or more owners are a Gnosis Safe (child safes). There can be
+multiple levels of nesting. The one parent safe that is not a child of any other safe is called a root safe.
+
+For a given transaction in a root safe within a nested safe setup, a scheduled transaction could be rejected by owners
+in child safes an arbitrary number of levels away. The `cancellation_threshold(safe)` of a child safe must be
+considered before registering that the child safe is rejecting the transaction.
+
+#### Severity: Medium
+
+If this invariant is broken, the cancellation flow would be not operational for nested setups, making the timelock less
+useful.
+
+### iSS-009: The Signatures For A Cancelled Transaction Can Not Be Reused
+
+The signatures for a cancelled transaction can not be reused.
+
+#### Severity: Medium
+
+If this invariant is broken, a malicious user could spam the multisig with transactions that must be cancelled,
+causing an impact close to a liveness failure through operational overhead.
+
+### iSS-010: Only Owners Can Signal Rejection Of Transactions
+
+To avoid rejection spamming, it must be verified upon rejection that the rejecting owner is an owner in the safe
+scheduled to execute the transaction, or in one of its child safes.
+
+#### Severity: Low
+
+If this invariant is broken, the event history of the timelock would contain useless events.
+
+## Function Specification
+
+### `configureLivenessModule`
+
+Configure the contract as a liveness module by setting the `liveness_challenge_period` and `fallback_owner`.
+
+- MUST allow an arbitrary number of `safe` contracts to use the contract as a module.
+- The contract MUST be enabled as a module on the `safe`.
+- MUST set the caller as a `safe`.
+- MUST take `liveness_challenge_period` and `fallback_owner` as parameters and store them as related to the `safe`.
+- If a challenge exists, it MUST be canceled, including emitting the appropriate events.
+- MUST emit a `ModuleConfigured` event with at least `liveness_challenge_period` and `fallback_owner` as parameters.
+
+### `clearLivenessModule`
+
+Removes the liveness module configuration by a previously enabled `safe`.
+
+- The contract MUST NOT be enabled as a module on the `safe`.
+- MUST erase the existing `liveness_challenge_period` and `fallback_owner` data related to the calling `safe`.
+- If a challenge exists, it MUST be cancelled, including emitting the appropriate events.
+- MUST emit a `ModuleCleared` event.
+
+### `configureTimelockGuard`
+
+Configure the contract as a timelock guard by setting the `timelock_delay`.
```

Comment:
Note that you can change the delay on the timelock, and scheduled transactions will be affected as a result.

Reply:
Oh, I'm a fool. You can only change the delay through a transaction that is in itself subject to the timelock delay, so at most you could make transactions that are executable already not executable by setting a longer delay, but not the opposite.

We can change the logic to record scheduling time instead of execution time. No preference here from me.

Reply:
There's no preference for me either here.